drgn_tools/meminfo.py

```diff
@@ -0,0 +1,767 @@
+        )
+        stats["ShmemPmdMapped"] = (
+            get_global_page_state(prog, "NR_SHMEM_PMDMAPPED") * unit
+        )
+        try:
+            stats["FileHugePages"] = (
+                get_global_page_state(prog, "NR_FILE_THPS") * unit
+            )
+        except LookupError:
+            stats["FileHugePages"] = -1
+        try:
+            stats["FilePmdMapped"] = (
+                get_global_page_state(prog, "NR_FILE_PMDMAPPED") * unit
+            )
+        except LookupError:
+            stats["FilePmdMapped"] = -1
+    else:
+        stats["AnonHugePages"] = (
+            get_global_page_state(prog, "NR_ANON_TRANSPARENT_HUGEPAGES") * unit
+        )
+        stats["ShmemHugePages"] = -1
+        stats["ShmemPmdMapped"] = -1
+        stats["FileHugePages"] = -1
+        stats["FilePmdMapped"] = -1
+
+    stats["CmaTotal"] = prog["totalcma_pages"].value_()
+    stats["CmaFree"] = get_global_page_state(prog, "NR_FREE_CMA_PAGES")
+    return stats
+
+
+def show_all_meminfo(prog: Program) -> None:
+    """
+    Dump various details about the memory subsystem.
+    This function must parse machine info to determine arch-specific parameters
+    before parsing all memory statistics.
+    """
+    stats = get_all_meminfo(prog)
+
+    # Ignore highmem stats as the target kernels do not have no highmem zones.
+    basic_meminfo_items = [
+        "MemTotal",
+        "MemFree",
+        "MemAvailable",
+        "Buffers",
+        "Cached",
+        "SwapCached",
+        "Active",
+        "Inactive",
+        "Active(anon)",
+        "Inactive(anon)",
+        "Active(file)",
+        "Inactive(file)",
+        "Unevictable",
+        "Mlocked",
+        "SwapTotal",
+        "SwapFree",
+        "Dirty",
+        "Writeback",
+        "AnonPages",
+        "Mapped",
+        "Shmem",
+        "KReclaimable",
+        "Slab",
+        "SReclaimable",
+        "SUnreclaim",
+        "KernelStack",
+        "PageTables",
+        "NFS_Unstable",
+        "Bounce",
+        "WritebackTmp",
+        "CommitLimit",
+        "Committed_AS",
+        "VmallocTotal",
+        "VmallocUsed",
+        "VmallocChunk",
+        "Percpu",
+        "HardwareCorrupted",
+    ]
+    hugepage_meminfo_items = [
+        "AnonHugePages",
+        "ShmemHugePages",
+        "ShmemPmdMapped",
+        "FileHugePages",
+        "FilePmdMapped",
+    ]
+    cma_meminfo_items = ["CmaTotal", "CmaFree"]
+
+    # Output
+    for item in basic_meminfo_items:
+        if item in ["KernelStack", "VmallocTotal", "HardwareCorrupted"]:
+            # These statistics items are in kB.
+            if item == "HardwareCorrupted":
+                print("HardwareCorrupted: %5d kB" % (stats[item]))
+            else:
+                print(
+                    "%s%s kB"
+                    % (
+                        (item + ":").ljust(16, " "),
+                        str(stats[item]).rjust(8, " "),
+                    )
+                )
```

Comment:
I'm not certain I understand the need for the two cases? Also, I think you could shorten this code quite a bit by using f-strings. The following might not be _exactly_ equivalent to your code, but it's a similar idea:

```python
print(f"{item + ':':<15} {stats[item]: 8d} kB")
```

Reply:
This is to keep the original spacing as done in the kernel implementation.
"HardwareCorrupted:" has a length 18 and exceeds the target text 15 by 3, so the statistics number's length is reduced from 8 to 5.

I added a comment to explain this and also unified methods of formatting strings in this meminfo module.